libwacom/libwacom-database.c

```diff
@@ -266,25 +266,38 @@ libwacom_parse_stylus_keyfile(WacomDeviceDatabase *db, const char *path)
 		stylus->eraser_type = eraser_type_from_str (type);
 		g_free (type);
 
-		if (stylus->eraser_type == WACOM_ERASER_NONE) {
-			stylus->has_eraser = g_key_file_get_boolean(keyfile, groups[i], "HasEraser", &error);
-			if (error && error->code == G_KEY_FILE_ERROR_INVALID_VALUE)
-				g_warning ("Stylus %s (%s) %s\n", stylus->name, groups[i], error->message);
-			g_clear_error (&error);
-			stylus->has_lens = g_key_file_get_boolean(keyfile, groups[i], "HasLens", &error);
-			if (error && error->code == G_KEY_FILE_ERROR_INVALID_VALUE)
-				g_warning ("Stylus %s (%s) %s\n", stylus->name, groups[i], error->message);
-			g_clear_error (&error);
-			stylus->has_wheel = g_key_file_get_boolean(keyfile, groups[i], "HasWheel", &error);
-			if (error && error->code == G_KEY_FILE_ERROR_INVALID_VALUE)
-				g_warning ("Stylus %s (%s) %s\n", stylus->name, groups[i], error->message);
-			g_clear_error (&error);
-		} else {
-			stylus->has_eraser = FALSE;
-			stylus->has_lens = FALSE;
-			stylus->has_wheel = FALSE;
+		string_list = g_key_file_get_string_list (keyfile, groups[i], "PairedIds", NULL, NULL);
+		if (string_list) {
+			GArray *array;
+			guint i;
+
+			array = g_array_new (FALSE, FALSE, sizeof(int));
+			for (i = 0; string_list[i]; i++) {
+				char *endptr;
+				long val;
+
+				val = strtol (string_list[i], &endptr, 0);
```

Comment:
probably best to copy `safe_atoi` from libinput for this? There's also `g_ascii_string_to_unsigned()` but it doesn't handle hex properly.

Reply:
Looks like `safe_atoi_base` could handle the hex IDs, though with the minor caveat at the IDs would then have to always be specified in hex. I suppose we could also add "0" to the list of accepted bases, but I don't think we get much from allowing both decimal and hex.

Reply:
I would cheer on any CI patches that check for the pen IDs to always be in `0xabcd` format. No point having this part-decimal, part-hex anyway.

Reply:
New commit added for this.